darts/models/torch_forecasting_model.py

```diff
@@ -541,71 +563,108 @@ def predict_from_dataset(self,
                     delayed(self._build_forecast_series)(
                         batch_prediction[batch_idx],
                         input_series_dataset[dataset_idx][0]
                     )
-                    for batch_idx, dataset_idx in enumerate(indices)
+                    for batch_idx, dataset_idx in enumerate(batch_indices)
                 )
 
                 predictions.extend(ts_forecasts)
 
         return predictions
 
+    def _predict_batch_block_model(self,
+                                   n,
+                                   input_series,
+                                   cov_future,
+                                   roll_size) -> Sequence[TimeSeries]:
+
+        batch_prediction = []
+        out = self.model(input_series)[:, self.first_prediction_index:, :]
+        batch_prediction.append(out[:, :roll_size, :])
+        prediction_length = roll_size
+
+        while prediction_length < n:
+
+            # roll over input series to contain latest target and covariate
+            input_series = torch.roll(input_series, -roll_size, 1)
+
+            # update target input to include next `roll_size` predictions
+            if self.input_chunk_length >= roll_size:
+                input_series[:, -roll_size:, :self.output_dim] = out[:, :roll_size, :]
+            else:
+                input_series[:, :, :self.output_dim] = out[:, -self.input_chunk_length:, :]
+
+            # update covariates to include next `roll_size` predictions into the future
+            if cov_future is not None and self.input_chunk_length >= roll_size:
+                input_series[:, -roll_size:, self.output_dim:] = (
+                    cov_future[:, prediction_length - roll_size:prediction_length, :]
+                )
+            elif cov_future is not None:
+                input_series[:, :, self.output_dim:] = (
+                    cov_future[:, prediction_length - self.input_chunk_length:prediction_length, :]
+                )
+
+            # take only last part of the output sequence where needed
+            out = self.model(input_series)[:, self.first_prediction_index:, :]
+
+            # update predictions depending on how many data points have been predicted
+            if prediction_length <= n - self.output_chunk_length - roll_size:
+                batch_prediction.append(out[:, :roll_size, :])
+                prediction_length += roll_size
+            elif prediction_length < n - self.output_chunk_length:
+                # if we produce have `n - output_chunk_length < #predictions < n` we want to only use
+                # the predictions and covariates necessary to exactly reach `n - output_chunk_length`,
+                # so that the final forecast produces exactly the right number of predictions to reach `n`
+                spillover_prediction_length = (prediction_length + roll_size) - (n - self.output_chunk_length)
+                roll_size -= spillover_prediction_length
```

Comment:
`roll_size` cannot become negative here? (I guess it works, just asking)

Reply:
It cannot become negative since `spillover_prediction_length < roll_size`. Proof:
```
n - self.output_chunk_length > prediction_length (precondition of elif block)
```
implies
```
0 > (prediction_length) - (n - self.output_chunk_length)
```
implies
```
roll_size > (prediction_length + roll_size) - (n - self.output_chunk_length) = spillover_prediction_length
```

Reply:
Beautiful, thanks!